dowhy/gcm/_noise.py

```diff
@@ -20,3 +40,108 @@ def compute_noise_from_data(causal_model: InvertibleStructuralCausalModel, obser
+    :param approx_prediction_model: Prediction model for approximating f. The model is trained based on drawn noise and
+                                    target samples.
+    :param num_training_samples: Number of drawn samples for training the predictor based on approx_predictor_factory.
+                                 Note: This parameter is ignored if approx_predictor_factory is None.
+    :return: A tuple, where the first value is a callable (the function f) that expects a numpy array X as input and
+    the second value is a list with nodes that represents the expected order of the columns in X.
+    """
+    validate_causal_dag(causal_model.graph)
+
+    if approx_prediction_model is not None:
+        return _get_approx_noise_dependent_function(
+            StructuralCausalModel(node_connected_subgraph_view(causal_model.graph, target_node)),
+            target_node,
+            approx_prediction_model,
+            num_training_samples)
+    else:
+        return _get_exact_noise_dependent_function(
+            StructuralCausalModel(node_connected_subgraph_view(causal_model.graph, target_node)),
+            target_node)
+
+
+def _get_exact_noise_dependent_function(causal_model: StructuralCausalModel,
+                                        target_node: Any) \
+        -> Tuple[Callable[[np.ndarray], np.ndarray], List[Any]]:
+    nodes_order = list(nx.topological_sort(causal_model.graph))
+
+    def predict_method(noise_samples: np.ndarray) -> np.ndarray:
+        return compute_data_from_noise(causal_model,
+                                       pd.DataFrame(noise_samples,
+                                                    columns=[x for x in nodes_order]))[target_node].to_numpy()
+
+    return predict_method, nodes_order
+
+
+def _get_approx_noise_dependent_function(causal_model: StructuralCausalModel,
+                                         target_node: Any,
+                                         approx_prediction_model: PredictionModel,
+                                         num_training_samples: int) \
+        -> Tuple[Callable[[np.ndarray], np.ndarray], List[Any]]:
+    nodes_order = list(nx.topological_sort(causal_model.graph))
+
+    node_samples, noise_samples = \
+        noise_samples_of_ancestors(causal_model, target_node, num_training_samples)
+
+    approx_prediction_model.fit(shape_into_2d(noise_samples[nodes_order].to_numpy()),
+                                shape_into_2d(node_samples[target_node].to_numpy()))
+
+    return approx_prediction_model.predict, nodes_order
+
+
+def noise_samples_of_ancestors(causal_model: StructuralCausalModel,
+                               target_node: Any,
+                               num_samples: int) -> Tuple[pd.DataFrame, pd.DataFrame]:
+    sorted_nodes = nx.topological_sort(causal_model.graph)
+    all_ancestors_of_node = nx.ancestors(causal_model.graph, target_node)
+    all_ancestors_of_node.update({target_node})
```

Comment:
Yes. Since the noise of the target node is also an input. For instance: ```X -> Y -> Z``` with ```Y := f(X, N_Y)```, ```Z := g(Y, N_Z)``` then we would represent Z as: ```Z := g(f(X, N_Y), N_Z) = h(X, N_Y, N_Z)```, i.e. ```N_Z``` here is an input as well.